source/common/config/delta_subscription_state.cc

```diff
@@ -53,6 +62,28 @@ void DeltaSubscriptionState::updateSubscriptionInterest(
     names_added_.erase(r);
     names_removed_.insert(r);
   }
+  // If we unsubscribe from wildcard resource, drop all the resources that came from wildcard from
+  // cache.
+  if (cur_removed.contains(Wildcard)) {
+    wildcard_resource_state_.clear();
+  }
+  // Check if this is a legacy wildcard subscription request. If we repeatedly call this function
+  // with empty cur_added and cur_removed, we keep the legacy wildcard subscription.
+  if (is_legacy_wildcard_) {
+    is_legacy_wildcard_ = cur_added.empty() && cur_removed.empty();
+  } else {
+    // This is a legacy wildcard subscription if we have not expressed interest in any resources so
```

Comment:
Does this only happen on the first request? Or can it reset the legacy wildcard status on stream reconnect?

Reply:
Hm, turns out to be a tricky question.

So at first I thought that there should be a chance of going back to legacy wildcard subscription in scenario like:

1. On stream establishment, we send an empty request (legacy wildcard subscription)
2. The we subscribe to some a resource named `r` (so we technically switch to explicit wildcard subscription).
3. Some time later we unsubscribe from the resource `r`, so at this point we are left with the wildcard subscription only.
4. Stream failure happens, so we re-establish the connection. Should we send the empty request (legacy wildcard subscription) or a request that subscribes to `*` (explicit wildcard subscription)?

I tried to achieve sending the empty request, but the code actually will send `*` explicitly (because `requested_resource_state_` will already contain `*`, so it won't be empty, thus making `is_legacy_wildcard_` false in https://github.com/envoyproxy/envoy/blob/db884fe1732a0a1b76f1c6caf064b474e1265b5d/source/common/config/delta_subscription_state.cc#L77-L78). I could try to fix the detection (and add more tests checking it).

But, now imagine another case:

1. On stream establishment, we send an empty request (legacy wildcard subscription)
2. The we subscribe to some a resource named `r` (so we technically switch to explicit wildcard subscription).
3. Some time later we unsubscribe from the resource `r`, so at this point we are left with the wildcard subscription only.
4. Some time later we unsubscribe from the wildcard resource. So we are subscribed to nothing at all.
5. Stream failure happens, so we re-establish the connection. What request we send? It can't be empty, because this will be a legacy wildcard request. I suppose we shouldn't be sending any requests now.

Reply:
> So at first I thought that there should be a chance of going back to legacy wildcard subscription in scenario like:
> 
> 1. On stream establishment, we send an empty request (legacy wildcard subscription)
> 2. The we subscribe to some a resource named `r` (so we technically switch to explicit wildcard subscription).
> 3. Some time later we unsubscribe from the resource `r`, so at this point we are left with the wildcard subscription only.
> 4. Stream failure happens, so we re-establish the connection. Should we send the empty request (legacy wildcard subscription) or a request that subscribes to `*` (explicit wildcard subscription)?

When the stream fails, the state of having previously subscribed to a non-wildcard resource on the stream should reset to false.  So when the new stream starts, if the only subscription is the wildcard subscription, I think the right behavior is to send the empty resource names list (legacy subscription).

> But, now imagine another case:
> 
> 1. On stream establishment, we send an empty request (legacy wildcard subscription)
> 2. The we subscribe to some a resource named `r` (so we technically switch to explicit wildcard subscription).
> 3. Some time later we unsubscribe from the resource `r`, so at this point we are left with the wildcard subscription only.
> 4. Some time later we unsubscribe from the wildcard resource. So we are subscribed to nothing at all.
> 5. Stream failure happens, so we re-establish the connection. What request we send? It can't be empty, because this will be a legacy wildcard request. I suppose we shouldn't be sending any requests now.

If we are not subscribed to any resources for a given resource type, then when the stream is re-established, we shouldn't send *any* request for that resource type.